cumulus/pallets/parachain-system/src/validate_block/trie_recorder.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus. If not, see <http://www.gnu.org/licenses/>.
+
+//! Provide a specialized trie-recorder and provider for use in validate-block.
+//!
+//! This file defines two main structs, [`SizeOnlyRecorder`] and
+//! [`SizeOnlyRecorderProvider`]. They are used to track the current
+//! proof-size without actually recording the accessed nodes themselves.
+
+use codec::Encode;
+
+use sp_std::{
+	cell::{RefCell, RefMut},
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	rc::Rc,
+};
+use sp_trie::{NodeCodec, ProofSizeProvider, StorageProof};
+use trie_db::{Hasher, RecordedForKey, TrieAccess};
+
+/// A trie recorder that only keeps track of the proof size.
+///
+/// The internal size counting logic should align
+/// with ['sp_trie::recorder::Recorder'].
+pub(crate) struct SizeOnlyRecorder<'a, H: Hasher> {
+	seen_nodes: RefMut<'a, BTreeSet<H::Out>>,
+	encoded_size: RefMut<'a, usize>,
+	recorded_keys: RefMut<'a, BTreeMap<Rc<[u8]>, RecordedForKey>>,
+}
+
+impl<'a, H: trie_db::Hasher> trie_db::TrieRecorder<H::Out> for SizeOnlyRecorder<'a, H> {
+	fn record(&mut self, access: TrieAccess<'_, H::Out>) {
+		let mut encoded_size_update = 0;
+		match access {
+			TrieAccess::NodeOwned { hash, node_owned } =>
+				if self.seen_nodes.insert(hash) {
+					let node = node_owned.to_encoded::<NodeCodec<H>>();
+					encoded_size_update += node.encoded_size();
```

Comment:
if node.encoded_size() >= 32 && not_root {
encoded_size_update -= CHILDREN_COMPACT_SAVE_SIZE;
}

(with children compact save size being ~ 32 I would need to redo the calcualation, but basically compacting proof will remove the parent node hash (32 or 33 if hash size encoded I don't remember) and replace it by a 0 length inline node (1byte)).
not_root is not that necessary to be honest just starting the initial size at CHILDREN_COMPACT_SAVE_SIZE might be a better idea.
I think same logic should apply for EncodedNode access.

Reply:
The logic that lives here regarding the proof size needs to have the exact same size counting behaviour as the  [reference implementation](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/trie/src/recorder.rs#L305). In this version I was not planning to account for any kind of compation or compression.

Reply:
The fact that cumulus uses compaction would make this better as default I think (calculating it as a second pass is costly (need to parse all the proof trie).

But yes you are right,  also the reference implementation should change, doing it in a follow up pr would make sense  (as long as the host function don t expose this behavior in an actual runtime, to avoid versioning to a new version) and would be fairly straight forward:
changes from these comments, same on reference impl, then fix existing test and propbably extend testing recorder size in  test like `recorder_equivalence_no_cache` to also check against the size of the actual proof.